src/assets/scripts/client/aircraft/AircraftModel.js

```diff
@@ -1806,32 +1805,29 @@ export default class AircraftModel {
      * @return {number}
      */
     _calculateTargetedAltitudeVnav() {
-        const hardRestrictedWaypointModel = this.fms.nextHardAltitudeRestrictedWaypoint;
-
-        switch (this.flightPhase) {
-            case FLIGHT_PHASE.TAKEOFF:
-            case FLIGHT_PHASE.CLIMB:
-                if (_isNil(hardRestrictedWaypointModel)) {
-                    return;
-                }
+        const nextAltitudeMaximumWaypoint = this.fms.getNextWaypointWithMaximumAltitudeAtOrBelow(this.altitude);
+        const nextAltitudeMinimumWaypoint = this.fms.getNextWaypointWithMinimumAltitudeAtOrAbove(this.altitude);
+        const maximumAltitudeExists = !_isNil(nextAltitudeMaximumWaypoint);
+        const minimumAltitudeExists = !_isNil(nextAltitudeMinimumWaypoint);
 
-                return this._calculateTargetedAltitudeVnavClimb(hardRestrictedWaypointModel);
+        if (!maximumAltitudeExists && !minimumAltitudeExists) {
+            return this.mcp.altitude;
+        }
 
-            case FLIGHT_PHASE.CRUISE: {
-                if (!this.shouldStartDescent()) {
-                    return;
-                }
+        if (maximumAltitudeExists && minimumAltitudeExists) {
```

Comment:
just a few lines up you check for: 
```js
!maximumAltitudeExists && !minimumAltitudeExists
```

if that case is true, you return.  so the check here on this line seems repetitive.  Then again, you know how I feel about successive conditional blocks in the same method.

Reply:
Yeah, I do know what you mean. I don't see a way to resolve it though, but as far as the branching...
- NEITHER minimum nor maximum
- BOTH minimum and maximum
- Minimum BUT NOT maximum
- Maximum BUT NOT minimum
...all of these require distinctly different actions, which are contained within their own methods which are called from here. Open to ideas to rearrange, but I don't see anything myself.

Reply:
fair enough.  sometimes the logic is gross because the logic _is_ gross